Move the development workflow from the wiki to the documentation

asmeurer opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->


This moves the document https://github.com/sympy/sympy/wiki/Development-workflow to the contributing section of the documentation.

So far, I have just copied it over verbatim, but I plan to do a lot of refactoring and cleaning up of the guide as well. So if anyone has any comments on things I should try to address as I go through this, please let me know. 

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- other
  - Move the development workflow page from the wiki to the contributing section of the documentation.
<!-- END RELEASE NOTES -->

asmeurer committed:
Add the development-workflow.rst file from the wiki

asmeurer committed:
Fix RST errors in the development workflow file

asmeurer commented:
I'd like to propose a top-level organization for the contributing section of the docs. Here is a draft outline.

- Introduction to Contributing
- Contributing Guide
  - Setting up Development Environment
  - Git and GitHub Instructions
  - Code Conventions
  - Running the Tests
- Contributing to Documentation
  - Documentation Style Guide
  - Building the Documentation
- Policies
  - Deprecation Policy
  - [Issue Policy](https://github.com/sympy/sympy/wiki/Issue-Policy)
  - [Pull Request Policy](https://github.com/sympy/sympy/wiki/Pull-Request-Policy)
  - [Python Version Support Policy](https://github.com/sympy/sympy/wiki/Python-version-support-policy)
- Miscellaneous
  - Debugging Tips
  - List of Dependencies

This is all based on the content that currently exists in the [contributing section of the docs](https://docs.sympy.org/dev/contributing/index.html), the [introduction to contributing](https://github.com/sympy/sympy/wiki/Introduction-to-contributing) and [development workflow](https://github.com/sympy/sympy/wiki/Development-workflow) pages on the wiki, and the linked policy pages on the wiki. Some of these would be further split into subpages (especially the git/GitHub instructions section since there's quite a bit of content currently written for that).

This is my first draft of a layout, so suggestions for improvements are welcome. 

I'm not sure about the name for the miscellaneous section. It would include various pages containing development tips and other documents that are useful for developers (like the list of dependencies). A lot of other stuff that's currently on the wiki could also go here, although much of it would need to be cleaned up first. Basically, anything that would be useful to contributors only (if it's useful to a general SymPy user, it would go in the other main sections of the documentation).

oscarbenjamin commented:
I think it's important to have a simple page somewhere that summarises how to contribute a PR concisely for the benefit of someone who already knows how to use git and github etc and has already contributed to other projects. Basically it should just list the minimum information that is needed concisely like below. Any more than minimal routinely needed detail on any topic should just link to another page.

Another topic I would suggest to add is "writing tests". I would also like to add a section on using pytest to run the tests (but I can do that myself later).

## Setup dev environment

1. Install relevant Python version
1. Fork and clone sympy
2. cd into your local repo
3. pip install mpmath (and usually gmpy2)
4. pip install numpy, matplotlib, ipython etc if those are needed
5. pip install --editable .

## Make a PR

1. Make a PR branch (don't use master)
2. Make changes and test if working.
3. Run tests (but usually not the whole test suite) or build docs etc.
4. Add tests for anything fixed or added (usually in adjacent `test_` file).
6. Can run `flake8` or `bin/test quality` but these are also handled in CI.
7. Commit changes
8. For a first contribution add name to .mailmap then run `bin/mailmap_check.py` and commit changes to .mailmap.
9. Push to your fork and open a PR.

## Troubleshooting PR

Often there will be failures listed in CI or requests for changes to the PR. You can always push more commits to the branch (no need to open a new PR):

1. Release note - use expected headings and don't just delete the release note bit from PR template. Can still edit the release note after opening the PR.
1. flake8 type problems - probably need to remove some imports or something.
2. bin/test quality - need to strip whitespace
3. commits add or delete files - if unintended then commits should be squashed or just start again with a new branch and PR.
4. .mailmap problems - if first contribution make sure to add yourself to .mailmap and run the script. Check the actual name and email address of your commits with `git log`. Have you already been added to the AUTHORS file but changed your git config since?
5. Test failures - try running those particular tests locally. Check if the test itself is correct and consider whether the changed behaviour is correct. Sometimes a test needs to be changed just because precise form of output has changed. Sometimes the failure shows a bug in the changes made in the PR.
6. Doc build failures - try building the docs yourself. Check carefully for warnings which are usually buried in the sphinx output.
7. Usually all jobs in CI need to pass but sometimes there are jobs that routinely fail e.g. if they test unreleased upstream branches of CPython, NumPy, SciPy etc. or if a new version of Sphinx etc breaks something.

asmeurer commented:
So maybe we can have a separate "Contributing Guide for People Who Have Contributed to Open Source Before" guide. that is really short and to the point. 

I want to organize what is already written first, but I will stub out new content to be written.

> Doc build failures - try building the docs yourself. Check carefully for warnings which are usually buried in the sphinx output.

With https://github.com/sympy/sympy/pull/23695, the sphinx warnings are printed again at the very end of the build, and they are always printed in color on CI, so this is much better than it used to be.

bertiewooster commented:
Seems good. Suggest having a section on doctests, both writing and running them. That's relevant to both code (should have examples as doctests in docstrings) and documentation (often want to have code examples in non-API documentation), so not sure where to put that section...

bertiewooster commented:
Actually, suggest making Contributing Guide general to any type of contributing (code or documentation) by moving

- Code Conventions
- Running the Tests

into a new, subsequent section called Contributing to Code or the like. That way, everyone will read Contributing Guide, and people can choose to read Contributing to Code, or Contributing to Documentation, or both, depending on their interest.

asmeurer committed:
Fix some more restructured text formatting in the development workflow

asmeurer committed:
Convert the development workflow to Markdown

asmeurer committed:
Some cleanups in the development workflow document

asmeurer committed:
Rename docs preview GitHub Action to lowercase

Hopefully this will make it sort to the top of the list

asmeurer committed:
More improvements to the development workflow document

asmeurer committed:
More cleanups to the development workflow document

asmeurer committed:
Fix a cross-reference

asmeurer committed:
Add Introduction to Contributing from the wiki

asmeurer committed:
Clean up the Introduction to Contributing document

asmeurer committed:
Start reorganizing the development workflow into separate guides

asmeurer committed:
Convert dev-setup from RST to Markdown

asmeurer committed:
Start cleaning up dev-setup

asmeurer committed:
More work on the development workflow docs

oscarbenjamin commented:
See also #24308

asmeurer committed:
Fix a typo

asmeurer committed:
Start writing a guide on writing tests

asmeurer committed:
Add the writing tests guide to the index

asmeurer committed:
Add sections for doctests, code quality, and updating existing tests

asmeurer committed:
Add some more of the writing tests document

asmeurer committed:
Write more stuff in the writing tests document

asmeurer commented:
I've started a guide on writing tests at https://github.com/sympy/sympy/pull/23879/files#diff-7315bf5583c24ad4cf3af7e463383865180cf79ac52d21761a06f49544f8b2b2. There is still some more to be written, especially about how to run the tests, and I still need to figure out a cleaner organization. But feel free to review what I have so far. 

I am also still in the process of writing/reorganizing the other development documentation, so some other bits of this PR are still in an unfinished state.

asmeurer committed:
Add a short section on running the tests

asmeurer committed:
Some fixes to the writing tests guide

asmeurer committed:
Add a note about dummy_eq to the writing tests guide

asmeurer committed:
Add a tests style guide to the writing tests guide

asmeurer committed:
Some updates to the doctests section

asmeurer committed:
Add a bit about generating a coverage report

asmeurer committed:
Don't make the index of the contributing guide hidden

asmeurer committed:
Remove "SymPy" from the page title

asmeurer committed:
Move the commiting guide to the workflow process page

asmeurer committed:
Use myst_linkify_fuzzy_links = False

asmeurer committed:
Add a "make a pull request" section to the workflow process page

asmeurer committed:
Remove the development-workflow page

All the useful content on that page has been moved to the new contributors
guide.

asmeurer committed:
Update the link to the mailmap instructions

asmeurer committed:
Use a better page title for the build-docs page

asmeurer committed:
Revert "Rename docs preview GitHub Action to lowercase"

This reverts commit 781720d844f3f0b46b465a682575334d51e5387a.

This doesn't actually affect the sorting. There doesn't seem to be a way to
actually do that.

asmeurer committed:
Merge branch 'master' into development-workflow

asmeurer committed:
Remove a broken cross-reference

asmeurer commented:
I'm marking this as ready for review. I've come to realize full scope of moving all development related pages from the wiki over is way too large for a single pull request, not to mention all the new content that can be written. 

I think if we can agree with the overall layout of what I've got here, then new content can be written/moved over later. I suggest using the docs preview in the checks below for reference.

The basic layout here is

- The top-level contributing section in the docs is the front page for any documentation related to contributing (this already exists in the [docs now](https://docs.sympy.org/latest/contributing/index.html)).
- The first page there is "introduction to contributing". This is based on https://github.com/sympy/sympy/wiki/introduction-to-contributing#get-familiar-using-the-software. It's a very short page with just the links to the important pages like the issue tracker, mailing list, and the various guides.
- After that there is the "guide for new contributors". As suggested, this is broken into two parts:
  - The "Setup Development Environment" page is designed for people who are new to Git and GitHub. It just lists the basic one-time steps to setup the dev environment.
  - The "Development Workflow Process" page assumes the reader is already familiar with the basics of using git and GitHub, and only details the things that are specific to contributing to SymPy.
  - I've omitted a lot of the git specific stuff from the development workflow page, like the guide to rebasing and merging. Honestly, I'm not even sure if we really need the "setup development environment" page which explains how to install git (the only SymPy specific thing on that page is "setup a virtual environment", which amounts to just "install mpmath"). When these guides were written, GitHub was still very new, even people who weren't new to coding might be new to using git. Nowadays, every open source project uses git and GitHub, so the number of people who are completely new to it is much smaller than it used to be. And there are now tons of guides on using both, including many which are much better written than ours. So I don't know if we really need to maintain half-baked guides on how to do various things in git, unless they are specific things to SymPy (like updating .mailmap). 
  - As requested at https://github.com/sympy/sympy/issues/24308, I've written a new guide just for writing tests. This covers all the testing topics I could think of. Please give special attention to [this guide](https://github.com/sympy/sympy/pull/23879/files#diff-7315bf5583c24ad4cf3af7e463383865180cf79ac52d21761a06f49544f8b2b2) when reviewing as almost all the content there is new.
- The old guides that duplicated a lot of this content have been removed. We should also considering removing the corresponding pages from the wiki once this is merged.

Ideally, many other pages from the wiki will also be moved over. Basically anything that is some sort of documentation for users or a policy page should go in the docs instead of the wiki. But this can be done later. The important thing is if the general layout here looks good, so that we don't have to change it in the future (and break links).

asmeurer committed:
Move the building docs guide to the guide for new contributors

asmeurer committed:
Remove the "make a contribution" page

This just pointed to the wiki guide, which is now in the docs themselves.

asmeurer committed:
Re-add build-docs guide

asmeurer committed:
Fix build-docs redirect

asmeurer committed:
Fix docs build errors

asmeurer committed:
Fix doctests

oscarbenjamin commented on doc/src/contributing/new-contributors-guide/writing-tests.md:
```diff
@@ -0,0 +1,1078 @@
+will run only the tests for the solvers.
+
+If you want, you can also use `pytest` to run the tests instead of the
+`./bin/test` tool.
```
We should give an example of usage like `pytest -m 'not slow' sympy/solvers`.

Perhaps there could be a separate page describing how to use pytest. It has several features that are very useful with sympy such as e.g. `-n4` for using multiple cores and `--lf` to rerun only tests that failed in the previous run.

oscarbenjamin commented on doc/src/contributing/new-contributors-guide/writing-tests.md:
````diff
@@ -0,0 +1,1078 @@
+./bin/doctest
+```
+
+to run the doctests. Note that the full test suite can take some time to run,
+so you may prefer to just run a subset of the tests, e.g., corresponding to
+the module you modified. You can do this by passing the name of the submodules
+or tests files to the test command. For example,
````
I think "you may prefer" is a bit of an understatement. Unless I am making some deep changes I very rarely run the entire test suite and almost always just run the tests for a specific module or even submodule. The full test suite is far too slow to use during development iteration. I would rather explain this the other way around i.e. first suggest `bin/test solvers` and then only afterwards suggest that in some situations you might want to run the whole test suite.

oscarbenjamin commented on doc/src/contributing/new-contributors-guide/writing-tests.md:
```diff
@@ -0,0 +1,1078 @@
+Another option is to just push your code up to GitHub and let the tests run on
+the CI. The GitHub Actions CI will run all the tests. However, it can take
+some time to finish, so it is usually advisable to run at least the basic
+tests before committing to avoid having to wait.
```
Perhaps an example here of how to interpret the output from CI and then run the test locally e.g. if you see this:
```
_____________________________________________________________________________________________________
_________________ sympy/printing/pretty/tests/test_pretty.py:test_upretty_sub_super _________________
Traceback (most recent call last):
  File "/home/oscar/current/sympy/sympy.git/sympy/printing/pretty/tests/test_pretty.py", line 317, in test_upretty_sub_super
    assert upretty( Symbol('beta_1_2') ) == 'β₁₂'
AssertionError
```
then you can reproduce the failure locally with (note the extra colon):
```
$ pytest sympy/printing/pretty/tests/test_pretty.py::test_upretty_sub_super
...
______________________________________ test_upretty_sub_super _______________________________________

    def test_upretty_sub_super():
>       assert upretty( Symbol('beta_1_2') ) == 'β₁₂'
E       AssertionError: assert 'β₁ ₂' == 'β₁₂'
E         - β₁₂
E         + β₁ ₂
E         ?   +

sympy/printing/pretty/tests/test_pretty.py:317: AssertionError
```

asmeurer replied:
It's a simple fix to make the SymPy test runner both accept and produce pytest-style double colon test specifications, so I've made that change here.

oscarbenjamin commented:
The writing tests guide looks good.

The development workflow process is an improvement. It's still too long for what I was thinking though. I really want a page for drive-by contributors who generally know what they are doing and just want to submit a quick PR. They just need a quick checklist to look at and do not need to be told how to use git etc. Lowering the barrier for experienced contributors is just as important as lowering the barrier for novices (if not more so).

asmeurer commented:
A checklist is a good idea. I've already done something similar at https://docs.sympy.org/latest/contributing/deprecations.html#checklist. 

I can add that as a new page (or maybe at the top), with links to more details on the longer page.

asmeurer committed:
Use the tasklist MyST extension for checkboxes instead of raw HTML

This lets you add checkbox lists to Markdown with

- [ ]

syntax (similar to GitHub).

This makes the checkboxes uncheckable for now, but this will be fixed once an
upstream MyST change is merged.

asmeurer committed:
Make tasklist checkbox items format like bullet points

asmeurer committed:
Whitespace cleanup

asmeurer committed:
Add a checklist for the development workflow

Also make some cross-reference names more consistent.

asmeurer commented:
Added a simple checklist to the top of the page. Let me know what you think.

asmeurer committed:
Make ./bin/test produce and accept pytest-style path::test_name specifications

asmeurer committed:
Add a cross-reference

asmeurer committed:
Add a section on debugging CI test failures

asmeurer committed:
Add checkboxes for docstrings and doctests

asmeurer committed:
Address review comments

asmeurer commented:
Addressed the review comments too.

oscarbenjamin commented on doc/src/contributing/new-contributors-guide/writing-tests.md:
````diff
@@ -85,13 +86,86 @@ or tests files to the test command. For example,
+The bit in between `_________________` is the name of the test. You can
+reproduce the test locally by copying and pasting this:
+
+```
+./bin/test sympy/printing/pretty/tests/test_pretty.py:test_upretty_sub_super
+```
+
+or
+
+```
+pytest sympy/printing/pretty/tests/test_pretty.py:test_upretty_sub_super
+```
````
Should these not be double colon?

asmeurer committed:
Fix docs build error

asmeurer committed:
Fix test example

asmeurer committed:
Fix bug in test collection

oscarbenjamin commented:
Thechecklist looks good to me.

asmeurer commented on doc/src/contributing/introduction-to-contributing.md:
```diff
@@ -0,0 +1,105 @@
+
+## Join the Gitter chat room
+
+We have a live chat room where we discuss things about SymPy at Gitter: https://gitter.im/sympy/sympy. Here you may find other SymPy developers and you can ask questions in real-time.
```
I can omit it from here for now. We should probably have a bigger discussion about Gitter on the mailing list. If none of the core developers are using it we might consider sunsetting it, or at least adding a warning to it.

asmeurer committed:
Merge branch 'master' into development-workflow

asmeurer committed:
Remove mentions of Gitter from the dev guide

We need to discuss Gitter usage in the community before we decide to recommend
it.

Also fix some links.

asmeurer commented:
Any more comments on this?

oscarbenjamin commented on doc/src/contributing/new-contributors-guide/dev-setup.md:
```diff
@@ -0,0 +1,228 @@
+See <https://git-scm.com/book/sv/v2/Customizing-Git-Git-Configuration> for
+some more common git configuration options.
+
+### (Optional) Tune Your Bash Prompt for Git
+
+*This step is not required, but it can make working with git on the command
+line easier, so it is recommended.*
```
I would move or remove this whole section. It is a distraction compared to the important things like "open a github account" which are further down.

asmeurer replied:
I removed it. It is definitely something I recommend to new git users, but I'm also starting to lean towards we should have little to no very intro git stuff in our dev guide. So including on the very basics is a decent compromise for now.

oscarbenjamin replied:
I think that this is useful information but it would just be better on a separate "git tips" page. Right here there shouldn't be anything more than a link to that page though because none of this is really needed and it's important not to overwhelm people with unnecessary things.

oscarbenjamin commented on doc/src/contributing/new-contributors-guide/dev-setup.md:
````diff
@@ -0,0 +1,228 @@
+
+Then, on your machine browse to where you would like to store SymPy, and clone (download) the latest code from SymPy's original repository (about 77 MiB):
+
+```bash
+$ git clone https://github.com/sympy/sympy
+```
+
+Then assign your read-and-write repo to a remote called "github" (replace
+`<your-github-username>` with your GitHub username):
+
+```
+git remote add github git@github.com:<your-github-username>/sympy.git
+```
````
The origin remote will already be configured by `git clone` or at least it would if `git clone` used the `git@` rather than `https` url. Lower down the github remote is not mentioned. I would remove this "github" remote part.

asmeurer replied:
This is the fork remote. The example below this definitely should say "github" and not "upstream" in the output though.

oscarbenjamin replied:
Oh, wait this is suggesting an odd setup. Usually "origin" is the fork and "upstream" is the main repo (at least that's how I've always done it). I would say that the contributor should clone their fork and then add upstream as a remote.

I'm now wondering which of these gives a better workflow...

oscarbenjamin replied:
I would try to stick with what is the common workflow and I think that using upstream for the main repo and origin for the fork is it e.g. see all answers in:
https://stackoverflow.com/questions/7244321/how-do-i-update-or-sync-a-forked-repository-on-github

asmeurer replied:
Our dev guide has always used the "clone the main repo/add your fork as a remote" workflow. There is also the opposite workflow which a lot of people use. I personally prefer the workflow that we've always suggested, because 

1) it matches what've already been recommending for a while
2) it's more uniform. With the "clone your fork" model, you have to fork literally every repo that you clone. And it doesn't even work for repos not on GitHub (which is admittedly rare these days. It also doesn't work for the situation where you push directly to a repo without forking it. Obviously these workflows aren't directly relevant for SymPy development but it's why I generally find this workflow simpler for general git usage.
3) Aesthetically it's a little more symmetric to just treat your own fork as "another upstream" (actually this guide calls the fork upstream `github` so that it can just use the same name everywhere, but it might be better to call it `yourusername` instead).

I've no idea which workflow is more common but I believe both are used, so I wouldn't really say there's an argument there.

oscarbenjamin replied:
> I've no idea which workflow is more common

I'm pretty sure that the clone-your-fork model is more common and that there is more documentation and advice for that setup. Of course there can be advantages in doing it the other way but I think it is better to keep this in line with wider practice.

It seems particularly disjointed to recommend this setup without giving any instructions for how to use it to do anything useful. The most common thing that someone would want to do is keep their local repo up to date either for testing or for making new PRs. There are instructions elsewhere for doing that (e.g. the SO post I linked above) but those are targeted at the more common clone-your-fork setup.

asmeurer replied:
The [original guide](https://github.com/sympy/sympy/wiki/Development-workflow) uses it a lot but I guess those all got cut. Honestly, I'm starting to wonder if we should just cut this entire page.

oscarbenjamin replied:
> I'm starting to wonder if we should just cut this entire page.

I think this page is good and has useful information although I would cut the "Configure Git Settings" part.

Otherwise the sections are

- Install Git
- Configure Your Name and Email in Git
- Setup GitHub
- Create a GitHub Account
- Setup SSH Keys
- Fork SymPy
- Get the SymPy Code
- Virtual Environment Setup

Those are all probably part of any "dev setup".

GitHub itself now seems to push access tokens rather than SSH keys. I don't know if there is any advantage there but otherwise everything above is essential information for contributors. Is this information duplicated elsewhere?

asmeurer replied:
Access tokens for pushing to git? I haven't seen that. How do you even set that up?

oscarbenjamin replied:
See here:
https://docs.github.com/en/authentication
https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations

Since GitHub stopped accepting username and password for https authentification their docs recommend personal access tokens by default. All of their docs suggest https URLs that can only be used with personal access tokens. They continue to allow ssh but I wouldn't be surprised if they decide against those at some point.

For now though I still use ssh and I suppose you do too and we can continue to recommend that in the sympy dev guide.

asmeurer replied:
I'm not sure that they are actually recommending to use a token instead of SSH for git. Tokens are used as a replacement for username and password, but SSH is also an option. Basically, instead of authenticating through git with your GitHub password, you now have to generate a special password token to use instead, which can have limited scope compared to your GitHub password.

I'm not a security expert, so maybe my read on this is wrong, but as far as I know, SSH is still more secure than using a token. It only requires uploading a public key to GitHub, so a breach of GitHub could not leak your private key. Personally I use [Secretive](https://github.com/maxgoedjen/secretive) which stores my SSH key in my Mac's secure enclave which makes it impossible for the private key to leak at all. None of this is possible with a token, which is basically just a special password to your account. The only advantage of tokens over SSH keys is that they can have restricted access and can automatically expire. But SSH keys already have some degree of restricted access. They can only be used to push to a repo, whereas tokens can potentially be used for any API access. You could in principle regularly rotate your SSH keys, although I'm not sure how many people do that. GitHub doesn't require it, so it must not be a huge threat.

oscarbenjamin replied:
> tokens can potentially be used for any API access

You set the access when creating the token.

asmeurer replied:
Going back to the original point, though, as you say, these things are all part of any dev setup. So why do we need to document it specially in our docs? I think we could just as well find a decent guide someone else has made and point to it. That reduces the maintenance burden for us. And honestly, this guide has always been only decent at best, whereas I'm sure there are much better written guides out there. I expect this is the sort of thing the [Scientific Python](https://discuss.scientific-python.org/) project is interested in doing, if it hasn't been done already. 

The only SymPy specific thing in this entire guide is the bit about installing "mpmath" as a dependency. That can easily be moved as a small section in the other page. 

This git guide stuff was a lot more relevant when it was started a long time ago, because back then git was still quite new and even seasoned developers hadn't used it yet. But now, the only people who don't know how to use Git and GitHub are people who are new to open source development in general. This is a lot of people, to be sure, but this applies to basically all of open source. There are literally hundreds of guides like this one out there, or varying quality. It's shouldn't hard to find a decent one that we can recommend.

oscarbenjamin replied:
There can be still some benefit in duplicating information from elsewhere if it makes the situation clearer for users.

oscarbenjamin commented on doc/src/contributing/new-contributors-guide/dev-setup.md:
````diff
@@ -0,0 +1,228 @@
+
+You may want to take advantage of using virtual environments to isolate your development version of SymPy from any system wide installed versions, e.g. from `apt-get install python-sympy`.
+
+We recommend using `conda` to create a virtual environment:
+
+```bash
+$ conda create -n sympy-dev python=3 mpmath flake8
+```
````
I don't think that this should "recommend" conda since there is no real reason to use conda for sympy itself (although it can be useful for some optional dependencies).

A simple setup using pip and venv is like:
```
cd sympy
python -m venv .venv
source .venv/bin/activate
pip install -e .
```
Then when opening a new terminal for development only the `cd` and `source` steps are needed.

asmeurer replied:
Starting with conda is a good idea for other reasons, but I can change this to show both ways.

oscarbenjamin replied:
Consider someone who already knows how to use virtual environments and pip (or whatever alternatives) and already knows what conda is but doesn't usually use it. There is no sympy-specific reason for that person to use conda so we should not give the impression that contributing to sympy is a reason for using conda.

oscarbenjamin replied:
I see you've changed it already

asmeurer committed:
Address some review comments for the development workflow

oscarbenjamin commented on doc/src/contributing/new-contributors-guide/workflow-process.md:
```diff
@@ -0,0 +1,708 @@
+  description field. It can be edited at any time before the pull request is
+  merged.
+
+- [ ] **Respond to review comments.** All SymPy pull requests must be reviewed
+  by someone else before they can be merged.
+
+
+## Pick an issue to fix
+
+The best way to start with the main code base is to fix some existing bugs.
+Peruse the ["Easy to fix"
+issues](https://github.com/sympy/sympy/issues?q=is%3Aopen+is%3Aissue+label%3A%22Easy+to+Fix%22)
```
The checklist above is good because it is nice and short for someone who already knows what they are doing in terms of git, github, python etc. That hypothetical person wants a quick list of what is needed that is partcular to sympy rather than say another project that they could contribute to via github. I think it is good to have a nice short page for that hypothetical contributor where it is assumed that the contributor already knows of an issue to be fixed and wants a low-friction interaction to submit a PR because they know how to do that.

From this point on though it moves into information that is not so relevant for someone who already knows what they are doing e.g. "don't commit to master" which I think does not belong in the same page.

oscarbenjamin replied:
Of course ending this page with a bunch of links like "how to write tests", "how to build the docs" etc is good but this page should be kept short so that someone who knows what they are doing can read it entirely in a short time.

oscarbenjamin replied:
Note that other places refer to this page as the one to read for someone who generally knows what they are doing so the hypothetical contributor has been directed here.

oscarbenjamin replied:
Or maybe it's fine if the section above just ends with something like "that's all that is needed but below is some more information about various aspects of contributing and some links for quick navigation to the sections below.

Basically I want someone who knows how to improve the code and how to make a pull request not to have to waste a lot of time figuring out what to do. Asking that person to read long pages full of irrelevant stuff discourages contribution.

asmeurer replied:
A surprising number of people don't know that you shouldn't commit to master. There's actually quite a skill gap. In fact, if you polled SymPy contributors on how comfortable they are with git/GitHub I'd wager the results would skew toward low self-reported skill, even among people who actually do know a lot.

I guess the question here is whether this is better on this page or on a separate page. I don't really feel strongly either way. I ended up putting it on the same page because it felt better that way, as it makes it easier to cross-reference things on the same page. But it's really not a big deal. Hopefully it's clear that the checklist is just a summary of the sections below, but we can make it clearer if it isn't.

oscarbenjamin replied:
Maybe there could just be some text that clarifies that the rest of the document does not need to be read for someone who clearly understands how to follow the steps in the checklist.

oscarbenjamin replied:
Of course they can read it if they want but we make it clear that there are no unusual surprises down there for someone who knows what they are doing so they can skip it if it looks straight-forward.

asmeurer committed:
Small fixes to the writing tests document

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into tmp_merge

oscarbenjamin committed:
doc: remove old dev-setup.rst

(Fixing a botched merge)

oscarbenjamin commented:
Okay, let's get this in. We can iterate on improvements later but this is a big improvement to the dev docs.

logicalNil commented on sympy/testing/runtests.py:
```diff
@@ -2282,7 +2292,7 @@ def add_text(mytext):
                 filename, f, (t, val, tb) = e
                 self.write_center("", "_")
-                self.write_center("%s:%s" % (filename, f.__name__), "_")
+                self.write_center("%s::%s" % (filename, f.__name__), "_")
```
What New implementations Can we Get from this
